Framework/DataHandling/src/LoadILLDiffraction.cpp

```diff
@@ -192,8 +200,159 @@ void LoadILLDiffraction::loadMetaData() {
+ *should be the position of the first tube.
+ *
+ * @param instrumentWorkspace The empty workspace containing the instrument
+ * @return A V3D object containing the position of the relevant component
+ */
+V3D LoadILLDiffraction::getReferenceComponentPosition(
+    const MatrixWorkspace_sptr &instrumentWorkspace) {
+  if (m_instName == "D2B") {
+    return instrumentWorkspace->getInstrument()
+        ->getComponentByName("tube_1")
+        ->getPos();
+  }
+
+  const auto &detInfo = instrumentWorkspace->detectorInfo();
+  const auto &indexOfFirstDet = detInfo.indexOf(1);
+  return detInfo.position(indexOfFirstDet);
+}
+
+/**
+ * Convert from absolute rotation angle, around the sample, of tube 1, to a
+ *relative rotation angle around the sample.
+ *
+ * @param tubeRotations Input is the absolute rotations around the sample of
+ *tube 1, output is the relative rotations required from the IDF for all
+ *detectors
+ * @param firstTubePosition A V3D object containing the position of the first
+ *tube
+ */
+void LoadILLDiffraction::calculateRelativeRotations(
+    std::vector<double> &tubeRotations, const V3D &firstTubePosition) {
+  // The rotations in the NeXus file are the absolute rotation of the first
+  // tube. Here we get the angle of that tube as defined in the IDF.
+
+  const double firstTubeRotationAngle =
+      firstTubePosition.angle(V3D(0, 0, 1)) * rad2deg;
+  g_log.debug() << "First tube rotation:" << firstTubeRotationAngle << "\n";
+
+  // Now pass calculate the rotations to apply for each time index.
+  std::transform(
+      tubeRotations.begin(), tubeRotations.end(), tubeRotations.begin(),
+      [&](double angle) { return (angle - firstTubeRotationAngle); });
+
+  g_log.debug() << "Instrument rotations to be applied : "
+                << tubeRotations.front() << " to " << tubeRotations.back()
+                << "\n";
+}
+
+/**
+ * Fills the counts for the instrument with moving detectors.
+ *
+ * @param data : detector data
+ * @param scan : scan data
+ */
+void LoadILLDiffraction::fillMovingInstrumentScan(const NXUInt &data,
+                                                  const NXDouble &scan) {
+
+  std::vector<double> axis = {0.};
+  std::vector<double> monitor = getMonitor(scan);
+
+  // First load the monitors
+  for (size_t i = 0; i < NUMBER_MONITORS; ++i) {
+    for (size_t j = 0; j < m_numberScanPoints; ++j) {
+      m_outWorkspace->mutableY(j + i * m_numberScanPoints) = monitor[j];
+      m_outWorkspace->mutableE(j + i * m_numberScanPoints) = sqrt(monitor[j]);
+      m_outWorkspace->mutableX(j + i * m_numberScanPoints) = axis;
+    }
+  }
+
+  // Then load the detector spectra
+  for (size_t i = NUMBER_MONITORS;
+       i < m_numberDetectorsActual + NUMBER_MONITORS; ++i) {
+    for (size_t j = 0; j < m_numberScanPoints; ++j) {
+      unsigned int y = data(static_cast<int>(j),
+                            static_cast<int>((i - NUMBER_MONITORS) / 128),
```

Comment:
I think the NeXus file makes more sense (see changes I have just pushed). D20 provides a good justification, with the dead pixels, but even that aside it should be more reliable and easier to read the dimensions from NeXus. You could do it for D2B, but if you ever had the tubes split across different subgroups it would get tricky to count them.